lib/ssl/src/ssl_certificate.erl

```diff
@@ -518,37 +534,39 @@ maybe_shorten_path(Path, PartialChainHandler, Default) ->
             Default
     end.
 
-new_trusteded_path(DerCert, [DerCert | Chain], _) ->
-    {DerCert, Chain};
+new_trusteded_path(DerCert, [{DerCert,OTPCert} | Chain], _) ->
+    {{DerCert,OTPCert}, Chain};
 new_trusteded_path(DerCert, [_ | Rest], Default) ->
     new_trusteded_path(DerCert, Rest, Default);
 new_trusteded_path(_, [], Default) ->
     %% User did not pick a cert present 
     %% in the cert chain so ignore
     Default.
 
-handle_incomplete_chain([PeerCert| _] = Chain0, PartialChainHandler, Default, CertDbHandle, CertDbRef) ->
-    %% We received an incomplete chain, that is not all certs expected to be present are present. 
-    %% See if we have the certificates to rebuild it. 
-    case certificate_chain(PeerCert, CertDbHandle, CertDbRef) of
-        {ok, _, [PeerCert | _] = Chain} when Chain =/= Chain0 -> %% Chain candidate found          
-            case lists:prefix(Chain0, Chain) of 
+handle_incomplete_chain([{DerPeer, ErlPeer}=Peer| _] = Chain0, PartialChainHandler, Default, CertDbHandle, CertDbRef) ->
+    %% We received an incomplete chain, that is not all certs expected to be present are present.
+    %% See if we have the certificates to rebuild it.
+    case certificate_chain(ErlPeer, DerPeer, CertDbHandle, CertDbRef, [Peer], []) of
+        {ok, _, [Peer | _] = Chain} when Chain =/= Chain0 -> %% Chain candidate found
+            case lists:prefix(Chain0, Chain) of
                 true ->
-                    handle_partial_chain(lists:reverse(Chain), PartialChainHandler, CertDbHandle, CertDbRef);
+                    Res  = handle_partial_chain(lists:reverse(Chain), PartialChainHandler, CertDbHandle, CertDbRef),
+                    to_der(trusted_cert_and_paths, Res);
                 false ->
-                    Default
+                    to_der(trusted_cert_and_paths, Default)
             end;
         _  ->
-            Default
+            to_der(trusted_cert_and_paths, Default)
     end.
 
 extraneous_chains(Certs) ->
     %% If some certs claim to be the same cert that is have the same
     %% subject field we should create a list of possible chain certs
     %% for each such cert. Only one chain, if any, should be
     %% verifiable using available ROOT certs.
-    Subjects = [{subject(Cert), Cert} || Cert <- Certs],
+    Subjects = [{subject(OTP), Cert} || {_,OTP} = Cert <- Certs],
     Duplicates = find_duplicates(Subjects),
+    io:format("~P~n", [Duplicates, 20]),
```

Comment:
I think you forgot to remove a debug printout?!

Reply:
I think so too :-)